win32/src/call_once.c

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "windows.h"
+
+#include "azure_c_logging/xlogging.h"
+
+#include "azure_c_pal/interlocked.h"
+#include "azure_c_pal/interlocked_hl.h"
+
+#include "azure_c_pal/call_once.h"
+
+MU_DEFINE_ENUM_STRINGS(CALL_ONCE_RESULT, CALL_ONCE_RESULT_VALUES);
+
+/*returns CALL_ONCE_PROCEED if calling code should do proceed to call the code, any other value = code already called*/
+CALL_ONCE_RESULT call_once_begin(volatile_atomic int32_t* state)
+{
+    CALL_ONCE_RESULT result = CALL_ONCE_ALREADY_CALLED; /*optimistically initialize*/
+    
+    int32_t state_local;
+
+    /* 0 - (CALL_ONCE_NOT_CALLED) = not called*/
+    /* 1 - calling... */
+    /* 2 - already called*/
+
+    /*Codes_SRS_CALL_ONCE_02_001: [ call_once_begin shall use interlocked_compare_exchange(state, 1, 0) to determine if user has alredy indicated that the init code was executed with success. ]*/
+    /*Codes_SRS_CALL_ONCE_02_002: [ If interlocked_compare_exchange returns 2 then call_once_begin shall return CALL_ONCE_ALREADY_CALLED. ]*/
+    while ((state_local = interlocked_compare_exchange(state, 1, 0)) != 2)
+    {
+        if (state_local == 0) /*it was not initialized, now it is "1" (initializing)*/
+        {
+            /*Codes_SRS_CALL_ONCE_02_004: [ If interlocked_compare_exchange returns 0 then call_once_begin shall return CALL_ONCE_PROCEED. ]*/
+            /*only 1 thread can get here, (the current one) so letting it proceed to initialize*/
+            result = CALL_ONCE_PROCEED;
+            break;
+        }
+        else
+        {
+            /*Codes_SRS_CALL_ONCE_02_003: [ If interlocked_compare_exchange returns 1 then call_once_begin shall call InterlockedHL_WaitForNotValue(state, 1, INFINITE) and call again interlocked_compare_exchange(state, 1, 0). ]*/
+            /*state cannot be "2" because while condition, so that means state is "1", and that means, need to wait until it is not 1*/
+            if (InterlockedHL_WaitForNotValue((LONG*)state, 1, INFINITE) != INTERLOCKED_HL_OK)

```

Comment:
>InterlockedHL_WaitForNotValue [](start = 16, length = 29)

ahh this is probably why this is in win32 :( still hasn't been updated

Reply:
uhhuh.

---
In reply to: [466041928](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466041928) [](ancestors = 466041928)

Reply:
we are both wrong.

There is a sync_wait_all - availale for both linux and windows. So I will just do that.

---
In reply to: [466126700](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466126700) [](ancestors = 466126700,466041928)

Reply:
... myeah, except nobody did a "reals" for azure_c_pal/sync.g, so now I have to do that too! ./rake

---
In reply to: [466133040](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466133040) [](ancestors = 466133040,466126700,466041928)

Reply:
... while doing this I ran into a catch22, that I do not have the energy at this hour to solve it. Here it is:
1. gballoc_ll_win32heap really likes to use lazy_init  because it needs to create a heap (just one, just once)
2. call_once requires wait_on_adress (which is separately provided on linux/windows)

So the chain of dependencies is like gballoc_ll_win32heap(pal_win32)===requires===>lazy_init===requires===>call_once===requires===>wait_on_address. Wait on address is in pal_win32. Rather circular. Hmmmm. In my initial PR I had them all clumped together (that is one way of doing it). The other way is to have the dependency broken, where gballoc_ll_win32heap is NOT placed into pal_win32. It could be placed in a higher level lib, as it takes a dependency of "common". Sort of having pal_win32_ll, pal_win32_common (build on top of ll) and a pal_win32_hl - build on top of pal_win32_common. Those who build an rebuild have always something to build.


---
In reply to: [466139006](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466139006) [](ancestors = 466139006,466133040,466126700,466041928)

Reply:
So I've decided to take a deferred task to do this in a more proper way. Right now luckily linux doesn't need call_once_lazy_init but that will be the time to do it.

---
In reply to: [466152661](https://github.com/Azure/azure-c-pal/pull/28#discussion_r466152661) [](ancestors = 466152661,466139006,466133040,466126700,466041928)